python/uv/private/lock.bzl

```diff
@@ -13,113 +13,424 @@
+    },
+)
 
-def lock(*, name, srcs, out, upgrade = False, universal = True, args = [], **kwargs):
-    """Pin the requirements based on the src files.
+def _args(ctx):
+    """A small helper to ensure that the right args are pushed to the _RunLockInfo provider"""
+    run_info = []
+    args = ctx.actions.args()
+
+    def _add_args(arg, maybe_value = None):
+        run_info.append(arg)
+        if maybe_value:
+            args.add(arg, maybe_value)
+            run_info.append(maybe_value)
+        else:
+            args.add(arg)
+
+    def _add_all(name, all_args = None, **kwargs):
+        if not all_args and type(name) == "list":
+            all_args = name
+            name = None
+
+        before_each = kwargs.get("before_each")
+        if name:
+            args.add_all(name, all_args, **kwargs)
+            run_info.append(name)
+        else:
+            args.add_all(all_args, **kwargs)
+
+        for arg in all_args:
+            if before_each:
+                run_info.append(before_each)
+            run_info.append(arg)
+
+    return struct(
+        run_info = run_info,
+        run_shell = args,
+        add = _add_args,
+        add_all = _add_all,
+    )
+
+def _lock_impl(ctx):
+    srcs = ctx.files.srcs
+    output = ctx.actions.declare_file(ctx.label.name + ".out")
+
+    toolchain_info = ctx.toolchains[UV_TOOLCHAIN_TYPE]
+    uv = toolchain_info.uv_toolchain_info.uv[DefaultInfo].files_to_run.executable
 
-    Differences with the current {obj}`compile_pip_requirements` rule:
-    - This is implemented in shell and `uv`.
-    - This does not error out if the output file does not exist yet.
-    - Supports transitions out of the box.
-    - The execution of the lock file generation is happening inside of a build
-      action in a `genrule`.
+    py_runtime = ctx.toolchains[TARGET_TOOLCHAIN_TYPE].py3_runtime
+
+    args = _args(ctx)
+    args.add_all([
+        uv,
+        "pip",
+        "compile",
+        "--no-python-downloads",
+        "--no-cache",
+    ])
+    args.add("--custom-compile-command", "bazel run //{}:{}".format(
+        ctx.label.package,
+        ctx.attr.update_target,
+    ))
+    if ctx.attr.generate_hashes:
+        args.add("--generate-hashes")
+    if not ctx.attr.strip_extras:
+        args.add("--no-strip-extras")
+    args.add_all(ctx.files.build_constraints, before_each = "--build-constraints")
+    args.add_all(ctx.files.constraints, before_each = "--constraints")
+    args.add_all(ctx.attr.args)
+
+    python = py_runtime.interpreter if py_runtime.interpreter else py_runtime.interpreter_path
+    args.add("--python", python)
+    args.add_all(srcs)
+
+    args.run_shell.add("--output-file", output)
+
+    # These arguments does not change behaviour, but it reduces the output from
+    # the command, which is especially verbose in stderr.
+    args.run_shell.add("--no-progress")
+    args.run_shell.add("--quiet")
+
+    ctx.actions.run_shell(
```

Comment:
I'm having trouble understanding why this step needs the copy step as part of its execution.

Isn't this the same behavior?

```
srcs = list(ctx.attr.srcs)
if existing_file:
  srcs.append(existing_file)
output = declare_file(name + ".out")
ctx.actions.run([uv, "--output={output}"] + srcs, inputs=srcs, output=output)
```

uv is going to overwrite whatever --output specifies, right?

Reply:
`uv` expects the previous output to be in the location that is defined by `--output={output}`. If you pass it as a source then you will have extra log lines `via requirements-existing.txt`, which is not what you want here.